File: Python/leetcode_79.py
'''
Given an m x n grid of characters board and a string word,
return true if word exists in the grid.

The word can be constructed from letters of sequentially
adjacent cells, where adjacent cells are
horizontally or vertically neighboring. The same letter
cell may not be used more than once.
'''


def exist(board, word):
    rows, cols = len(board), len(board[0])

    def dfs(r, c, i):
        if i == len(word):  # All letters matched
            return True
        if r < 0 or c < 0 or r >= rows or c >= cols:
            return False
        if board[r][c] != word[i]:
            return False

        # Temporarily mark the cell as visited
        temp = board[r][c]
        board[r][c] = '#'

        # Explore all 4 directions
        found = (
            dfs(r+1, c, i+1) or
            dfs(r-1, c, i+1) or
            dfs(r, c+1, i+1) or
            dfs(r, c-1, i+1)
        )

        # Restore the cell
        board[r][c] = temp
        return found

    for r in range(rows):
        for c in range(cols):
            if dfs(r, c, 0):
                return True
    return False

File: JavaScript/leetcode_79.js
var exist = function (board, word) {
    const rows = board.length;
    const cols = board[0].length;

    function backtrack(r, c, i) {
        // If all characters are matched
        if (i === word.length) return true;

        // Check boundaries and current character
        if (
            r < 0 ||
            c < 0 ||
            r >= rows ||
            c >= cols ||
            board[r][c] !== word[i]
        ) {
            return false;
        }

        // Temporarily mark the cell as visited
        const temp = board[r][c];
        board[r][c] = "#";

        // Explore all 4 directions
        const found =
            backtrack(r + 1, c, i + 1) ||
            backtrack(r - 1, c, i + 1) ||
            backtrack(r, c + 1, i + 1) ||
            backtrack(r, c - 1, i + 1);

        // Restore the cell
        board[r][c] = temp;

        return found;
    }

    // Try to start from each cell
    for (let r = 0; r < rows; r++) {
        for (let c = 0; c < cols; c++) {
            if (backtrack(r, c, 0)) return true;
        }
    }

    return false;
};
